agent/consul/stream/event.go

```diff
@@ -101,6 +89,14 @@ func (p *PayloadEvents) HasReadPermission(authz acl.Authorizer) bool {
 	})
 }
 
+// TopicKey is required to satisfy the Payload interface but is not implemented
+// by PayloadEvents. PayloadEvents structs are constructed by Subscription.Next
+// *after* TopicKey has been used to dispatch the enclosed events to the correct
+// buffer.
```

Comment:
🤔 I wouldn't block merging this because of it, but I wonder if this is a sign that our Interface abstraction isn't expressing quite the right thing any more.

I wonder if it would be cleaner to have two interfaces - one for any event that is deliverable `Payload` and another for not-yet-routed events that embeds `Payload` but adds the `TopicKey` method, maybe `RoutablePayload` (I know, names). That way we don't need strange No-op implementations and we also encode the intent of each type in the type system - these types are really only for use internally once events have been routed to a buffer etc..

@dnephin what do you think as you designed the original interface here?

Reply:
I agree that having to implement methods on `Payload` types that will never have those methods called is unfortunate.  If I remember correctly that problem already existed when I extracted this interface, but extracting the Payload interface did make it more obvious.

I'd probably call it `KeyedPayload` (instead of `RoutablePayload`). If we were to do that I guess we have two options, either:
1. different types for the `[]Event` passed to `Publish` , and the event type we store internally and return to subscribers (ex: KeyedEvent vs Event), or
2. type-assert the `Payload` to `KeyedPayload` on Publish. 

I don't think we want to type-assert because we have no way of returning an error from `Publish`.

I think having a separate `Event` type to contain the different Payload types might be fine, but I'm not sure it's a big enough problem to expand the API the package in this way. If we were to do the suggestion I left below, and use a `type TopicKey string` then returning `""` from framing events doesn't seem too bad. A separate event type might be ok too.

Are there other options I've missed?

Reply:
+100. It definitely felt like the abstraction was beginning to creak a bit when adding those no-op methods. I'm going to play with this a bit and see if there are any other options 👷🏻

Reply:
Ok, I've given this a bit more thought. I'm wondering if it feels like more of a problem right now because the streaming API currently only covers health/connect events, and so there are just a lot more framing events than "normal" events.

It seems to me that there are three possible approaches here:

1. Keep things as-is (every `Payload` implements `TopicKey`)
   - **Pro:** Type system makes it impossible to forget to provide a key
   - **Con:** Interface feels a bit peculiar (all these no-op implementations)
   - **Pro:** Reflects the intention that framing events are the special case and most events _will_ have a key
2. Move the `TopicKey` method to another type (e.g. `KeyedPayload`)
   - **Pro:** Interfaces feel a bit cleaner, no no-op implementations
   - **Con:** Weakens type safety. We'd need to do type-assertions and panic or return an error at run-time (maybe generics would help here? 🤷)
3. Make `TopicKey` a struct field on `Event`
   - **Pro:** Similar to option 2 it makes the interfaces clearer
   - **Con:** Also similar to option 2, we can't help implementers to remember the key at compile-time

Overall, I think it might me sense to leave things as-is for now, and revisit in the future if we find ourselves adding lots of other methods to `Payload` that we need to no-op for framing events.

Reply:
Fine with leaving it as-is!